### PR TITLE
Fix balancer strategy

### PR DESCRIPTION
**Pull request by pjain1**
Fixes https://github.com/apache/druid/issues/10067, https://github.com/apache/druid/issues/10068 and https://github.com/apache/druid/issues/10069

### Description

Found multiple bugs in balancer strategy which are described in the issues above.

<hr>

This PR has:
- [x] been self-reviewed.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.

<hr>

##### Key changed/added classes in this PR
 * `CostBalancerStrategy`
 * `RandomBalancerStrategy`

### CODE REVIEW

File: server/src/main/java/org/apache/druid/server/coordinator/RandomBalancerStrategy.java [hunk 1]
```diff
@@ -28,20 +28,22 @@
 import java.util.NavigableSet;
 import java.util.Set;
 import java.util.concurrent.ThreadLocalRandom;
+import java.util.stream.Collectors;
 
 public class RandomBalancerStrategy implements BalancerStrategy
 {
   @Override
   public ServerHolder findNewSegmentHomeReplicator(DataSegment proposalSegment, List<ServerHolder> serverHolders)
```
Comment (nishantmonu51) [inline server/src/main/java/org/apache/druid/server/coordinator/RandomBalancerStrategy.java:36]: minor nit:mark return type as nullable

Comment (pjain1) [reply server/src/main/java/org/apache/druid/server/coordinator/RandomBalancerStrategy.java:36]: It is marked as Nullable in super class `BalancerStrategy`

**Comment by a2l007** [comment]
I feel we should mention this in the release notes since it modifies the default behavior for `druid.server.maxSize`. For clusters having historicals with this property set to zero, the coordinator will longer assign segments to them. I've added clarification to the docs via #10105

**Comment by clintropolis** [comment]
> I feel we should mention this in the release notes since it modifies the default behavior for druid.server.maxSize. For clusters having historicals with this property set to zero, the coordinator will longer assign segments to them. I've added clarification to the docs via #10105

Thanks for bringing this up, am working on draft of release notes right now and will be sure to add this. I don't think it is technically a change though, as far as I can tell it was always intended that the balancer uses the size historicals advertise in order to make segment placement decisions, just some flaws were causing the coordinator to assign them anyway I guess? I can't imagine the balancer was otherwise working if size was set to 0, since size is used pretty heavily to determine what can be moved and to where.

**Comment by a2l007** [comment]
I agree. However, the docs give an impression that it maybe okay to set this to the default value, although it isn't. Have tried to fix this in the linked PR.

**Pull request merged**